lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp

```diff
@@ -840,26 +842,51 @@ struct AIEObjectFifoStatefulTransformPass
             }
           }
 
-          if (found) {
-            int unrollFactor =
-                computeLCM(objFifoSizes); // also counts original loop body
-            // if loop iterations < unrollFactor, unroll the loop fully
-            if (forLoop.getSingleLowerBound() &&
-                forLoop.getSingleUpperBound() && forLoop.getSingleStep()) {
-              int64_t tripCount =
-                  constantTripCount(*(forLoop.getSingleLowerBound()),
-                                    *(forLoop.getSingleUpperBound()),
-                                    *(forLoop.getSingleStep()))
-                      .value_or(0);
-              if (tripCount < unrollFactor)
-                unrollFactor = tripCount;
-            }
-            if (failed(mlir::loopUnrollByFactor(forLoop, unrollFactor))) {
-              forLoop.emitOpError()
-                  << "could not be unrolled with unrollFactor: " << unrollFactor
-                  << "\n";
-              return WalkResult::interrupt();
-            }
+          int unrollFactor =
+              computeLCM(objFifoSizes); // also counts original loop body
+          Region *region = forLoop->getParentRegion();
+          while (remainder > 1 || found) {
+            region->walk([&](scf::ForOp remLoop) {
+              if (std::count(unrolledLoops.begin(), unrolledLoops.end(),
+                             remLoop) == 0) {
+                if (remLoop.getSingleLowerBound() &&
+                    remLoop.getSingleUpperBound() && remLoop.getSingleStep()) {
+                  int64_t tripCount =
+                      constantTripCount(*(remLoop.getSingleLowerBound()),
+                                        *(remLoop.getSingleUpperBound()),
+                                        *(remLoop.getSingleStep()))
+                          .value_or(0);
+                  // if loop iterations < unrollFactor, unroll the loop fully
+                  if (tripCount < unrollFactor)
+                    unrollFactor = tripCount;
+                  remainder = tripCount % unrollFactor;
```

Comment:
Adding @fifield's comment from the previous PR: this can trigger a division by zero when tripCount = 0. There should ideally be a test to check if constantTripCount() returned None that triggers an error message before then.